src/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,252 @@
+# See LICENSE file for licensing details.
+"""Generic kubernetes resource manager for gateway-api-integrator."""
+
+import abc
+import functools
+import logging
+import typing
+
+import lightkube
+import lightkube.generic_resource
+import lightkube.resources
+import lightkube.resources.apiextensions_v1
+import lightkube.resources.apps_v1
+import lightkube.resources.core_v1
+import lightkube.resources.discovery_v1
+from lightkube.core.exceptions import ApiError
+
+from resource_definition import GatewayResourceDefinition
+
+logger = logging.getLogger(__name__)
+
+AnyResource = typing.TypeVar(
+    "AnyResource",
+    lightkube.resources.core_v1.Endpoints,
+    lightkube.resources.discovery_v1.EndpointSlice,
+    lightkube.resources.core_v1.Service,
+    lightkube.resources.core_v1.Secret,
+    lightkube.generic_resource.GenericNamespacedResource,
+)
+
+ResourceDefinition: typing.TypeAlias = GatewayResourceDefinition
+
+CREATED_BY_LABEL = "gateway-api-integrator.charm.juju.is/managed-by"
+
+
+class InvalidResourceError(Exception):
+    """Custom error that indicates invalid resource definition.
+
+    Args:
+        msg: error message.
+    """
+
+    def __init__(self, msg: str):
+        """Construct the InvalidGatewayError object.
+
+        Args:
+            msg: error message.
+        """
+        self.msg = msg
+
+
+class KuberentesCreateResourceError(Exception):
+    """Custom error that indicates insufficient permission to create k8s resources.
+
+    Args:
+        msg: error message.
+    """
+
+    def __init__(self, msg: str):
+        """Construct the KuberentesCreateResourceError object.
+
+        Args:
+            msg: error message.
+        """
+        self.msg = msg
+
+
+def _map_k8s_auth_exception(func: typing.Callable) -> typing.Callable:
+    """Remap the kubernetes 403 ApiException to KuberentesCreateResourceError.
+
+    Args:
+        func: function to be wrapped.
+
+    Returns:
+        A wrapped function.
+    """
+
+    @functools.wraps(func)
+    def wrapper(*args: typing.Any, **kwargs: typing.Any) -> typing.Any:
+        """Remap the kubernetes 403 ApiException to KuberentesCreateResourceError.
+
+        Args:
+            args: function arguments.
+            kwargs: function keyword arguments.
+
+        Returns:
+            The function return value.
+
+        Raises:
+            ApiException: if the Python kubernetes raised an unknown ApiException
+            KuberentesCreateResourceError: if the Python kubernetes raised a permission error
+        """
+        try:
+            return func(*args, **kwargs)
+        except ApiError as exc:
+            if exc.status.code == 403:
+                logger.error(
+                    "Insufficient permissions to create the k8s service, "
+                    "will request `juju trust` to be run"
+                )
+                juju_trust_cmd = "juju trust <gateway-api-integrator> --scope=cluster"
+                raise KuberentesCreateResourceError(
```

Comment:
If the exception is raised only for 403 statuscode, couldn't be something more related like UnauthorizedAccessException or similar?

Reply:
good point, I was keeping it generic in case we'd want to use it for other 4xx return codes, but I changed the name to `InsufficientPermissionError`